brayns/engine/geometry/Geometry.h

```diff
@@ -20,190 +20,110 @@
+    Geometry(Type primitive)
+        : Geometry(std::vector<Type>{std::move(primitive)})
     {
-        if (primitives.size() >= std::numeric_limits<uint32_t>::max())
-        {
-            throw std::invalid_argument("Maximum number of primitives is 2^32 - 1");
-        }
     }
-};
 
-/**
- * @brief Utility class to commit geometry-specific parameters
- */
-template<typename T>
-class GeometryCommitter
-{
-public:
-    /**
-     * @brief Commits the type specific parameters
-     */
-    static void commit(const ospray::cpp::Geometry &osprayGeometry, const std::vector<T> &primitives)
-    {
-        (void)osprayGeometry;
-        (void)primitives;
-        throw std::runtime_error("Unhandled geometry type");
-    }
-};
+    Geometry(Geometry &&) noexcept = default;
+    Geometry &operator=(Geometry &&) noexcept = default;
 
-/**
- * @brief The Geometry class is a templated class to easily manage Ospray geometry at a high level.
- * It allows to add, retrieve and manipulate Geometry.
- */
-template<typename T>
-class Geometry
-{
-public:
-    Geometry(T primitive)
-        : _osprayGeometry(OsprayGeometryName<T>::get())
-    {
-        _primitives.push_back(std::move(primitive));
-        InputGeometryChecker<T>::check(_primitives);
-    }
-
-    Geometry(std::vector<T> primitives)
-        : _osprayGeometry(OsprayGeometryName<T>::get())
-        , _primitives(std::move(primitives))
-    {
-        assert(!_primitives.empty());
-        InputGeometryChecker<T>::check(_primitives);
-    }
+    Geometry(const Geometry &other);
+    Geometry &operator=(const Geometry &other);
 
     /**
-     * @brief Retrieves all geometry primitives in this Geometry object
+     * @brief Tries to cast the primitive data to the given type.
+     * @tparam Type primitive type to cast the data to
+     * @return const std::vector<Type>*, if the cast is successful, null otherwise
      */
-    const std::vector<T> &getPrimitives() const noexcept
+    template<typename Type>
+    const std::vector<Type> *as() const noexcept
     {
-        return _primitives;
+        if (auto cast = dynamic_cast<const GeometryData<Type> *>(_data.get()))
+        {
+            return &cast->primitives;
+        }
+        return nullptr;
     }
 
     /**
-     * @brief Allows to pass a callback to mainipulate all the geometries on thie Geometry buffer.
-     * The callback must have the signature void(const uint32_t index, GeometryType&).
+     * @copydoc Geometry::forEach<Callable>(Callable&&) const
      */
     template<typename Callable>
-    void forEach(const Callable &callback)
+    void forEach(Callable &&callback) noexcept
     {
-        _dirty = true;
-        const auto end = static_cast<uint32_t>(_primitives.size());
-        for (uint32_t i = 0; i < end; ++i)
+        using ArgType = DecayFirstArgType<Callable>; // typename ArgumentInferer<Callable>::argType;
+        auto cast = dynamic_cast<GeometryData<ArgType> *>(_data.get());
+        assert(cast);
```

Comment:
Remove comment ?

BTW if you want to throw if the cast fails you should use dynamic_cast<T &>.

And if the cast is not supposed to fail you can assert(dynamic_cast) and do a static_cast after so in release you don't have the cost of the RTTI (for micro-optimization :)).

Reply:
Addressed. Really cool trick.